src/decrypter.js

```diff
@@ -52,114 +50,138 @@ ntoh = function(word) {
 };
 
 /**
- * Schedule out an AES key for both encryption and decryption. This
- * is a low-level class. Use a cipher mode to do bulk encryption.
+ * Expand the S-box tables.
  *
- * @constructor
- * @param key {Array} The key as an array of 4, 6 or 8 words.
+ * @private
  */
-AES = function (key) {
-  this._precompute();
+const precompute = function() {
+  let _tables = [[[], [], [], [], []], [[], [], [], [], []]];
+  let encTable = _tables[0];
+  let decTable = _tables[1];
+  let sbox = encTable[4];
+  let sboxInv = decTable[4];
+  let i;
+  let x;
+  let xInv;
+  let d = [];
+  let th = [];
+  let x2;
+  let x4;
+  let x8;
+  let s;
+  let tEnc;
+  let tDec;
+
+  // Compute double and third tables
+  for (i = 0; i < 256; i++) {
+    th[(d[i] = i << 1 ^ (i >> 7) * 283) ^ i] = i;
+  }
+
+  for (x = xInv = 0; !sbox[x]; x ^= x2 || 1, xInv = th[xInv] || 1) {
+    // Compute sbox
+    s = xInv ^ xInv << 1 ^ xInv << 2 ^ xInv << 3 ^ xInv << 4;
+    s = s >> 8 ^ s & 255 ^ 99;
+    sbox[x] = s;
+    sboxInv[s] = x;
 
-  var i, j, tmp,
-    encKey, decKey,
-    sbox = this._tables[0][4], decTable = this._tables[1],
-    keyLen = key.length, rcon = 1;
+    // Compute MixColumns
+    x8 = d[x4 = d[x2 = d[x]]];
+    tDec = x8 * 0x1010101 ^ x4 * 0x10001 ^ x2 * 0x101 ^ x * 0x1010100;
+    tEnc = d[s] * 0x101 ^ s * 0x1010100;
 
-  if (keyLen !== 4 && keyLen !== 6 && keyLen !== 8) {
-    throw new Error("Invalid aes key size");
+    for (i = 0; i < 4; i++) {
+      encTable[i][x] = tEnc = tEnc << 24 ^ tEnc >>> 8;
+      decTable[i][s] = tDec = tDec << 24 ^ tDec >>> 8;
+    }
   }
 
-  encKey = key.slice(0);
-  decKey = [];
-  this._key = [encKey, decKey];
+  // Compactify. Considerable speedup on Firefox.
+  for (i = 0; i < 5; i++) {
+    encTable[i] = encTable[i].slice(0);
+    decTable[i] = decTable[i].slice(0);
+  }
+  return _tables;
+};
 
-  // schedule encryption keys
-  for (i = keyLen; i < 4 * keyLen + 28; i++) {
-    tmp = encKey[i-1];
+let aesTables;
 
-    // apply sbox
-    if (i%keyLen === 0 || (keyLen === 8 && i%keyLen === 4)) {
-      tmp = sbox[tmp>>>24]<<24 ^ sbox[tmp>>16&255]<<16 ^ sbox[tmp>>8&255]<<8 ^ sbox[tmp&255];
+/**
+ * Schedule out an AES key for both encryption and decryption. This
+ * is a low-level class. Use a cipher mode to do bulk encryption.
+ *
+ * @constructor
+ * @param key {Array} The key as an array of 4, 6 or 8 words.
+ */
+class AES {
+  constructor(key) {
+   /**
+    * The expanded S-box and inverse S-box tables. These will be computed
+    * on the client so that we don't have to send them down the wire.
+    *
+    * There are two tables, _tables[0] is for encryption and
+    * _tables[1] is for decryption.
+    *
+    * The first 4 sub-tables are the expanded S-box with MixColumns. The
+    * last (_tables[01][4]) is the S-box itself.
+    *
+    * @private
+    */
+    if (!aesTables) {
+      aesTables = precompute();
+    }
+    this._tables = JSON.parse(JSON.stringify(aesTables));
```

Comment:
Fancy :) Does this actually save us anything over recomputing on each construction?